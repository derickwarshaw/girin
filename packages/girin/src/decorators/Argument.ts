import { ArgumentMetadata } from '../metadata/ArgumentMetadata';
import { Generic } from '../metadata/Generic';


export interface ArgumentDecoratorOptions {
  defaultValue?: any;
  description?: string;
}

export function Argument(nameAndType: string, options?: ArgumentDecoratorOptions) {
  const [name, typeString] = nameAndType.split(/\s*:\s*/);

  return function(target: Object, fieldName: string, parameterIndex: number) {
    const definitionClass = target.constructor;
    ArgumentMetadata.create({
      definitionClass,
      fieldName,
      name,
      definedOrder: parameterIndex,
      generic: Generic.of(typeString),
      defaultValue: options && options.defaultValue,
      description: options && options.description,
    });
  }
}