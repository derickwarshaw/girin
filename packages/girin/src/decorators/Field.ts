import { FieldMetadata } from '../metadata/FieldMetadata';
import { Generic } from "../metadata/Generic";
import { globalMetadataStorage } from "../metadata/globalMetadataStorage";

export interface FieldDecoratorOptions {
  description?: string;
  deprecationReason?: string;
}

export function Field(type: string, options?: FieldDecoratorOptions): Function {
  return function(objectType: any, fieldName: string) {
    const definitionClass = objectType.constructor as any;
    FieldMetadata.create({
      fieldName,
      definitionClass,
      description: options && options.description,
      deprecationReason: options && options.deprecationReason,
      generic: Generic.of(type),
      meta: globalMetadataStorage,
    });
  };
}
