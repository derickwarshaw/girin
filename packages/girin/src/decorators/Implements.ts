import { MetadataStorage } from '../metadata/MetadataStorage';
import { ImplementsMetadata } from '../metadata/ImplementsMetadata';


export interface ImplementsDecoratorOptions {
  meta?: MetadataStorage;
}

export function Implements(targetNameOrDefinitionClass: Function | string, options: ImplementsDecoratorOptions = {}) {
  return function(definitionClass: Function & ImplementsDecoratorOptions) {
    const mergedConfig = Object.assign(options, definitionClass, { definitionClass, targetNameOrDefinitionClass });
    ImplementsMetadata.create(mergedConfig);
  };
}