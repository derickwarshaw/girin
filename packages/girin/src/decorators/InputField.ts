
import { InputFieldMetadata } from "../metadata/InputFieldMetadata";
import { MetadataStorage } from "../metadata/MetadataStorage";
import { Generic } from "../metadata/Generic";


export interface InputFieldDecoratorOptions {
  defaultValue?: any;
  description?: string;
  meta?: MetadataStorage;
}

export function InputField(nameAndType: string, options?: InputFieldDecoratorOptions): Function {
  const [ name, typeString ] = nameAndType.split(/\s*:\s*/);

  return function(definitionClass: Function, propertyName: undefined, parameterIndex: number) {
    InputFieldMetadata.create({
      definedOrder: parameterIndex,
      fieldName: name,
      generic: Generic.of(typeString),
      definitionClass: definitionClass,
      description: options && options.description,
      defaultValue: options && options.defaultValue,
      meta: options && options.meta,
    });
  };
}

