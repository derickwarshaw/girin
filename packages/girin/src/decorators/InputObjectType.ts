import { GraphQLInputFieldConfigMap, GraphQLInputObjectTypeConfig } from 'graphql';

import { MetadataStorage } from '../metadata/MetadataStorage';
import { InputObjectTypeMetadata } from '../metadata/InputObjectTypeMetadata';


export type InputObjectTypeDecoratorOptions = {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLInputFieldConfigMap;
  description?: string;
  astNode?: GraphQLInputObjectTypeConfig["astNode"];
}

export function InputObjectType(options: InputObjectTypeDecoratorOptions = {}) {
  return function(definitionClass: InputObjectTypeDecoratorOptions & Function) {
    const mergedConfig = Object.assign(options, definitionClass, { name: definitionClass.name, definitionClass });
    InputObjectTypeMetadata.create(mergedConfig);
  };
}
