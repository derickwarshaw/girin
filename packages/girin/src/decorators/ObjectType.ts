import { ObjectTypeMetadata } from '../metadata/ObjectTypeMetadata';
import { MetadataStorage } from '../metadata/MetadataStorage';
import { GraphQLObjectTypeConfig, GraphQLFieldConfigMap } from 'graphql';


export interface ObjectTypeDecoratorOptions {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLFieldConfigMap<any, any>;

  interfaces?: GraphQLObjectTypeConfig<any, any>["interfaces"];
  isTypeOf?: GraphQLObjectTypeConfig<any, any>["isTypeOf"];
  description?: GraphQLObjectTypeConfig<any, any>["description"];
  astNode?: GraphQLObjectTypeConfig<any, any>["astNode"];
  extensionASTNodes?: GraphQLObjectTypeConfig<any, any>["extensionASTNodes"];
}

export function ObjectType(options: ObjectTypeDecoratorOptions = {}) {
  return function(definitionClass: ObjectTypeDecoratorOptions & Function) {
    const mergedConfig = Object.assign(options, definitionClass, { name: definitionClass.name, definitionClass });
    ObjectTypeMetadata.create(mergedConfig);
  };
}
