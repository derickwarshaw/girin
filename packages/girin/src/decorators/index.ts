export { Argument } from './Argument';
export { Field } from './Field';
export { Implements } from './Implements';
export { InputField } from './InputField';
export { InputObjectType } from './InputObjectType';
export { ObjectType } from './ObjectType';
