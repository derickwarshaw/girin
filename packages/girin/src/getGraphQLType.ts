import { MetadataStorage } from "./metadata/MetadataStorage";
import { GraphQLType, isType } from 'graphql';

export interface TypeBuilderOptions {
  meta?: MetadataStorage;
}

export function getGraphQLType(definitionClassOrType: Function | GraphQLType, options?: TypeBuilderOptions): any {
  let meta: MetadataStorage;
  if (options && options.meta) {
    meta = options.meta;
  } else {
    meta = MetadataStorage.getMetadataStorage();
  }

  if (isType(definitionClassOrType)) {
    return definitionClassOrType;
  }

  const typeMetadata = meta.getDefinitionMetadata(definitionClassOrType);
  if (!typeMetadata) {
    throw new Error(`Given class ${definitionClassOrType.name} has no corresponding metadata`);
  }
  return typeMetadata.build.typeInstance;
}
