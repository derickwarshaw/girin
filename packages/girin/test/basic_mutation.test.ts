import { graphql, GraphQLSchema, printSchema } from 'graphql';

import { Field } from '../src/decorators/Field';
import { Argument } from '../src/decorators/Argument';
import { ObjectType } from '../src/decorators/ObjectType';
import { getGraphQLType } from '../src/getGraphQLType';

interface MemberSource {
  id: number;
  name: string;
  email: string;
}

@ObjectType()
class Member {
  constructor(source: MemberSource) {
    this.id = source.id;
    this.name = source.name;
    this.email = source.email;
  }
  @Field('Int!') id: number;
  @Field('String!') name: string;
  @Field('String!') email: string;
}

@ObjectType()
class Query {
  @Field('Member!')
  public getMember() {
    return new Member({ id: 1, name: 'Jonghyun', email: 'j@example.com' });
  }
}

@ObjectType()
class Mutation {
  @Field('Member!')
  public createMember(
    @Argument("name: String!") name: string,
    @Argument("email: String!") email: string,
  ) {
    return new Member({ id: 2, name, email });
  }
}

const schema = new GraphQLSchema({
  query: getGraphQLType(Query),
  mutation: getGraphQLType(Mutation),
});


describe('Basic mutation and schema generation', async () => {

  it('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('passes source and args to its resolver', async () => {
    const result = await graphql({ schema, source: `
      mutation {
        createMember(name: "Key" email: "k@example.com") {
          id
          name
          email
        }
      }
    `});
    expect(result.data!.createMember.id).toBe(2);
    expect(result.data!.createMember.name).toBe("Key");
    expect(result.data!.createMember.email).toBe("k@example.com");
  });

});
