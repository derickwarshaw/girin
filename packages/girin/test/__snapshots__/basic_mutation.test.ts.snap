// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic mutation and schema generation generates schema as expected 1`] = `
"type Member {
  id: Int!
  name: String!
  email: String!
}

type Mutation {
  createMember(email: String!, name: String!): Member!
}

type Query {
  getMember: Member!
}
"
`;
